Only proxy whitelisted request headers to ES server upstream

ycombinator opened the pull request:
Fixes #6484.

Besides the unit test included in this PR, I've tested this fix functionally in a couple of different ways:
- As described in https://github.com/elastic/kibana/issues/6484#issue-139643325, I set `http.cors.enabled: true` in the Elasticsearch config via esvm. Kibana loads up as expected.
- Looked at HTTP packets in Wireshark. There is no `Origin` header in the request that is sent to Elasticsearch:
  <img width="1552" alt="screen shot 2016-04-13 at 7 07 59 am" src="https://cloud.githubusercontent.com/assets/51061/14497411/c50a1cf8-014b-11e6-8822-16a1c02de58c.png">

w33ble commented:
I checked this out, it works, everything looks good, except for 1 thing. I think we need to be using a header whitelist instead of a blacklist like this is done currently. Additionally, that whitelist should be modifiable via the kibana.yml file, and then this PR can also close #6221

Seems like everyone else is in agreement on that, so I'm going to ask you to update this PR.

As for passing in a default whitelist, the theory goes that the only header that we *might* need to forward is `Authorization`, but even that probably isn't required; we should definitely get a broader consensus. Since we should lean on the side of more restriction by default, having an empty whitelist for now is probably sufficient.

ycombinator commented:
@w33ble I've updated the PR to use a whitelist approach as we discussed on Zoom and you summarized in your comment above. Using the default whitelist (no headers), here's what I see with Wireshark:

<img width="1552" alt="screen shot 2016-04-13 at 5 02 11 pm" src="https://cloud.githubusercontent.com/assets/51061/14513068/ac522fae-0199-11e6-8cb1-30f4c2139e8d.png">

I've also confirmed that the issue in #6484 goes away (because origin is not a whitelisted header by default). Interestingly, however, if I _do_ whitelist only `origin`, the issue in #6484 does not happen! So I experimented a bit more and found out that it is the combination of the `origin` and `user-agent` headers that causes the issue in #6484. Huh.

ycombinator commented:
With the configurable request header whitelist in place, this PR also resolves #6221

w33ble commented on src/plugins/elasticsearch/lib/map_uri.js:
```diff
@@ -1,6 +1,18 @@
+      return header.trim().toLowerCase();
+    });
+
+    return _.pick(originalHeaders, headersToKeepNormalized);
```
I'm not sure if `originalHeaders` will always come back lower-case or not. I think it's probably best to normalize them as well when filtering the list.

ycombinator replied:
Good call. Fixing.

w33ble commented:
Security doesn't work without the `authorization` header being passed. Auto-login via some proxy has been a really big ask lately as well - I think we should probably include that header in the default list.

w33ble commented:
Pending that one change, and our discussion about the default values tomorrow, this LGTM. 

I'd recommend passing to 1 more person for a second approval, since this is pretty important code. @panda01 is probably a decent candidate, since he's assigned to #6221

ycombinator commented on src/plugins/elasticsearch/lib/map_uri.js:
```diff
@@ -1,6 +1,23 @@
+  };
+
+  const filterHeaders = function (originalHeaders) {
+    const originalHeadersNormalized = _.mapKeys(originalHeaders, function (headerValue, headerName) {
```
BTW, I could replace `function (headerValue, headerName) { ... }` over here with `_.rearg(normalizeHeader, 1, 0)` and it would do the same thing. I chose not to use `_.rearg`, however, because I think it lowers readability of this code. I'm curious what the team norms are around this sort of decision: use more verbose code (like I have) to preserve readability and clear intent vs. use library functions to keep code succint. /cc @rashidkpc

panda01 replied:
I always err on the side of more verbose than less if i have to ask, but i think as a team, it's kinda up to people's styles.

panda01 commented:
The only thing I will say is this breaks security, before this gets merged i think the auth headers for shield definitely need to be added to the whitelist. Other than that, this LGTM!

ycombinator commented:
@skearns64 @tbragin @rashidkpc Can you weigh in here on what is a good default list of headers to proxy upstream to Elasticsearch? It sounds from @w33ble and @panda01's comments above that 'Authorization' should be on that list but I'd like consensus from you on whether:

1. the default list should be empty (current state of this PR), _or_
2. the default list should contain one item: the `Authorization` header, _or_
3. the default list should contain some other items: if so, which ones?

ycombinator commented:
Note to self: Talking to @epixa, this change (whitelist) needs to be made in `callWithRequest` too.

ycombinator commented:
After discussing this with @tanya and @epixa during today's meeting, I will modify the default whitelist to contain exactly one item: the `Authorization` header.

ycombinator commented:
@panda01 @epixa This is ready for review again. I've made two changes since the last review: 
- the default header whitelist now contains (only) `authorization`.
- updated `call_with_request.js` to use the whitelist instead of explicitly using the `authorization` header always.

ycombinator commented on src/plugins/elasticsearch/lib/expose_client.js:
```diff
@@ -78,8 +78,8 @@ module.exports = function (server) {
   server.expose('createClient', createClient);
-  server.expose('callWithRequestFactory', callWithRequest);
-  server.expose('callWithRequest', callWithRequest(noAuthClient));
+  server.expose('callWithRequestFactory', _.partial(callWithRequest, server));
```
@lukasolson Hey, I know you originally introduced `callWithRequestFactory` here in https://github.com/elastic/kibana/commit/5be340079b81cf934744159fc3aaa2af56d55747 so I was wondering if you could take a look at my refactor here. Let me know if you think there's a better way to do this.

BTW, I noticed that no code is actually using `callWithRequestFactory`. Is that expected?

lukasolson replied:
This refactor looks good to me. It is used here: https://github.com/elastic/x-plugins/blob/master/kibana/plugins/security/server/lib/get_client_shield.js#L11

ycombinator replied:
Thanks @lukasolson!

epixa commented on config/kibana.yml:
```diff
@@ -61,6 +61,11 @@
 # elasticsearch.requestTimeout: 30000
 
+# List of Kibana client-side headers to send to Elasticsearch. To send *no* client-side
+# headers, set this value to null, not []. Setting it to [] will send the default list
```
Commenting out the configuration is the standard way to use the default value for configurations, so I recommend ditching the extra logic for `null` altogether.  So `[]` sends no headers, commenting out sends the default (which is just authorization).

ycombinator replied:
So I tried to make it so `[]` sends no headers but I couldn't get that working with Joi. I think the issue is that it treats `[]` as empty and then replaces it with the default, which is `[ "authorization" ]`. That's why I had to play this silly game with null, which I don't like either.

epixa replied:
Hmm... that is frustrating, but I do think this is a burden we must overcome programmatically rather than placing it on the user.  @spalger do you have any idea how we can make joi allow empty arrays?

ycombinator replied:
Agreed re: solving this programmatically rather than passing on the burden to the user.

The issue isn't so much that Joi doesn't allow empty arrays — it does. Its more that we want it to see an empty array and treat it as-is, rather than replacing it with our specified default (which would be `[ 'authorization' ]`.

ycombinator replied:
What we really need is a way to know if the user explicitly set this value or not. If they did, we use it as-is; if they didn't we use our default. I don't know how to accomplish this with Joi.

spalger replied:
@ycombinator this was actually caused by a bug in read_yaml_config.js which was fixed yesterday by #6825

spalger replied:
If you pull master you should see that `elasticsearch.requestHeadersWhitelist: []` now clears the whitelist.

epixa commented on docs/kibana-yml.asciidoc:
```diff
@@ -29,7 +29,9 @@ to `false`.
 Elasticsearch. This value must be a positive integer.
-`elasticsearch.shardTimeout:`:: *Default: 0* Time in milliseconds for Elasticsearch to wait for responses from shards. Set 
+`elasticsearch.requestHeadersWhitelist:`:: *Default: `[ 'authorization' ]`* List of Kibana client-side headers to send to Elasticsearch.
+To send *no* client-side headers, set this value to null, not []. Setting it to [] will send the default list of headers.
```
Same as above.

epixa commented on docs/settings.asciidoc:
```diff
@@ -376,6 +376,10 @@ deprecated[4.2, The names of several Kibana server properties changed in the 4.2
 + 
 *default*: `500000`
 
+`elasticsearch.requestHeadersWhitelist:` added[5.0]:: List of Kibana client-side headers to send to Elasticsearch. To send *no* client-side headers, set this value to null, not []. Setting it to [] will send the default list of headers.
```
Goodness.. how many places is this configuration documented?

ycombinator commented:
@epixa Based on [Spencer's comment](https://github.com/elastic/kibana/pull/6896#discussion_r59914951), I've made some changes so that:

* If the user doesn't change their kibana.yml file at all, only the `authorization` header will be sent.
* If the user simply uncomments the setting line in kibana.yml, only the `authorization` header will be sent.
* If the user uncomments the setting line in kibana.yml and changes the value of the setting to nothing/blank or `[]` (i.e. empty list), no headers will be sent.

epixa commented on config/kibana.yml:
```diff
@@ -61,6 +61,11 @@
 
+# List of Kibana client-side headers to send to Elasticsearch. To send *no* client-side
+# headers, set this value to nothing/blank (i.e. elasticsearch.requestHeadersWhitelist: )
+# or the empty list (i.e. elasticsearch.requestHeadersWhitelist: [])
```
Let's give only one option for this: empty array.  Is there any situation where offering an alternative would be beneficial?

ycombinator replied:
Not particularly. I think as a user its intuitive for blank values to mean nothing (in addition to `[]` also meaning list with nothing in it) so I documented the alternative. I'm happy to only document the `[]` option (even though the blank would still work but might be less confusing to the user if we don't actually document it).

epixa replied:
If nothing else, for consistency sake I think we should stick to `[]`.  I'm sure there are other configurations in kibana.yml that technically would allow an empty value, but we don't advertise it as such.  Your call though.

ycombinator replied:
Sounds good to me. I'll document only `[]` here.

epixa commented on src/plugins/elasticsearch/index.js:
```diff
@@ -20,6 +22,7 @@ module.exports = function ({ Plugin }) {
         password: string(),
         shardTimeout: number().default(0),
         requestTimeout: number().default(30000),
+        requestHeadersWhitelist: array().items().single().default(DEFAULT_REQUEST_HEADERS),
```
What does `.items().single()` do here?

ycombinator replied:
It allows the user to define a single item array as a bare string. So they can define the setting as:

   elasticsearch.requestHeadersWhiteList: x-my-custom-header-1

instead of having to say:

   elasticsearch.requestHeadersWhiteList: [ x-my-custom-header-1 ]

`.single()` will convert `x-my-custom-header-1` to `[ x-my-custom-header-1 ]` because it knows the value of this setting must ultimately be an array.

epixa replied:
Got it, thanks.

epixa commented on src/plugins/elasticsearch/lib/create_proxy.js:
```diff
@@ -16,10 +16,11 @@ function createProxy(server, method, route, config) {
         mapUri: mapUri(server),
-        passThrough: true,
         agent: createAgent(server),
-        xforward: true,
```
I think we probably want to keep `xforward` enabled.  In this case, this is a header specifically injected by Kibana itself rather than a header sent by the user's browser.  This type of header only makes sense in the context of the direct proxy.

What do you think?

ycombinator replied:
I'm torn on this one. Its common for proxies to add `x-forwarded-*` headers and what we have here is indeed a proxy, so for that would be a reason to include these. On the other hand, `x-forwarded-for` will include information about the client (specifically, their IP address) and looking at comments such as https://github.com/elastic/kibana/issues/6221#issuecomment-183418042, I don't think we should include it.

I think, for now, we should err on the side of not including it. We can always add it in later if it is requested (either explicitly or via a config option) but it would be harder to take away later if we add it now and folks start relying on it somehow.

epixa replied:
I agree with the sentiment, but in this case, `xforward` is already enabled on the proxy, so we're actually talking about *removing* that feature without any means for people to turn it back on.

It seems like a different issue than this PR is addressing, so why don't we keep it on in this PR and address it in a separate issue/PR?  It'd probably be something that we would at least want to make configurable since there is no workaround for it if we just outright remove it.

ycombinator replied:
Addressing this in a separate issue/PR sounds good to me.

epixa commented on src/plugins/elasticsearch/lib/filter_headers.js:
```diff
@@ -0,0 +1,22 @@
+import _ from 'lodash';
+
+module.exports = function (originalHeaders, headersToKeep) {
```
Use `export default` here instead of `module.exports =`.  We're moving away from the commonjs module syntax in favor of the native module syntax.

ycombinator replied:
Will do!

ycombinator commented:
@epixa I've made the changes you suggested and this is ready for review again. Thanks!

epixa commented:
LGTM

ycombinator commented:
Labeling this as a breaking change because, after this change, we pass (by default) fewer headers upstream to Elasticsearch. So if there's a HTTP intermediary between Kibana and Elasticsearch relying on certain headers, they might break.
